### PR TITLE
Implemented Response as an object instead of a hash

### PR DESCRIPTION
**Pull request by Sacristan**
* Moved  ActiveMerchant::Billing::FirstDataGateway to ActiveMerchant::Billing::FirstData::Gateway
* Added ActiveMerchant::Billing::FirstData::Response -> now response is not just a hash, but object from what fields like transaction_id, parameters can be returned, also contains ok?, recurring? helper methods;
* Added result_text method to ActiveMerchant::Billing::FirstData::Response instance which returns description of result (not just code or OK, but an explanation to the status code);
* Updated Rspec to v 3.2;
* Rspec tests changed from should to expect style ;
* Added specs for ActiveMerchant::Billing::FirstData::Response ;
* ActiveMerchant::Billing::FirstData::Gateway.ssl_strict is to be set directly

### CODE REVIEW

**Comment by ebeigarts** [comment]
This is great :)

Can you make this backwards compatible by using  `FirstDataGateway` and `FirstDataResponse` classes and adding `Response#[]` method?

Also this kind of naming convention is used by `active_merchant` gem - `<name>Gateway` and `<name>Response`.

It would be nice if `FirstDataResponse` would inherit from [Response](https://github.com/Shopify/active_merchant/blob/master/lib/active_merchant/billing/response.rb) and implement `#success?`, `#authorization` (transaction_id), etc methods.

Global constants are bad, can you move them to `ActiveMerchant::Billing::FirstDataGateway::CURRENCY_CODES` and `ActiveMerchant::Billing::FirstDataResponse::FDL_RESPONSE_CODES`?

**Comment by Sacristan** [comment]
sure - ill update it ASAP

**Comment by Sacristan** [comment]
Im sorry - what You meant with "adding Response#[] method" ?

**Comment by ebeigarts** [comment]
```
class Response
  def [](value)
    @parameters[value]
  end
end
```

so we can access the values the old way using `response[:transaction_id]`

**Comment by Sacristan** [comment]
Got it! Thanks!

**Comment by Sacristan** [comment]
"
It would be nice if FirstDataResponse would inherit from Response and implement #success?, #authorization (transaction_id), etc methods.
"

authorization -> transaction_id ?

Current methods for ResponseObject
ok?
success? <- alias for ok?
recurring?
result
transaction_id
result_code
recc_pmnt_id
recc_pmnt_expiry
three_dee_secure
result_text
parameters

**Comment by Sacristan** [comment]
* Constants moved to their representable classes;
* Backwards compatibility ensured for Response and Gateway classes;
* Moved self.ssl_strict = false as it was in 1.2.0 (not configurable);
* Added authorization, success?, [], message, result_message methods to response.

**Comment by ebeigarts** [comment]
I think `FirstData::Gateway` is not needed if we have `FirstDataGateway`, can you move all the methods back?

**Comment by ebeigarts** [comment]
More comments :)

The `c` prefix is confusing in `c000: "Approved"`. `'000' => 'Approved'` would be better and less confusing.

I would prefer `Response#_3d_secure` over `#three_dee_secure`.

Why do you need that `FDL_` prefix? (`FDL_RESPONSE_CODES` => `RESPONSE_CODES`)

Since we now have `Response#success?`, we could remove `Response#ok?`.

Why do we need `Response#[]`? it should enough with `Response#[](value)`.

Inheriting `FirstDataResponse` from `ActiveMerchant::Billing::Response` would allow us
to reuse some functionality and be compatible with `ActiveMerchant::Billing::Response`.
There could be people that are using multiple gateways and expect gateways to return responses that respond to the exact same methods.

```ruby
class FirstDataResponse < Response
  def initialize(params, options = {})
    success == params[:result] == 'OK'
    message = FDL_RESPONSE_CODES[options[:result_code]]
    error_code = options[:result_code] unless success
    super success, message, params, options.merge(
      authorization: params[:transaction_id]
    )
  end

  [:transaction_id, :result_code, ...].each do |name|
    define_method name do
      params[name]
    end
  end

  ...
end
```

**Comment by Sacristan** [comment]
Please evaluate the code once again!

File: lib/active_merchant/billing/first_data_gateway.rb [hunk 1]
```diff
@@ -286,34 +284,33 @@ def redirect_url(trans_id = nil)
       end
 
       private
+        def lookup_currency(params)
+          params[:currency] = CURRENCY_CODES[params[:currency]] || params[:currency]
+        end
 
-      def lookup_currency(params)
-        params[:currency] = CURRENCY_CODES[params[:currency]] || params[:currency]
-      end
```
Comment (ebeigarts) [inline lib/active_merchant/billing/first_data_gateway.rb]: Can you leave the indention style as it was previously?

File: lib/active_merchant/billing/first_data_response.rb [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        unless success
+          error_code = options[:result_code]
+          error_message = message
+        end
+
+        super success, message, params, options.merge(
+          authorization: params[:transaction_id]
+        )
+
+        self.parameters=params
```
Comment (ebeigarts) [inline lib/active_merchant/billing/first_data_response.rb]: why do you need `self.parameters` if we already have `self.params`?

File: spec/active_merchant/billing/first_data/gateway_spec.rb [hunk 3]
```diff
@@ -42,130 +43,146 @@
...
+        )
+      else
+        @gateway = ActiveMerchant::Billing::FirstDataGateway.new(:pem => nil, :pem_password => nil)
+      end
+    end
+
+    it "should be present" do
+      expect(@gateway.present?).to be true
+    end
+  end
```
Comment (ebeigarts) [inline spec/active_merchant/billing/first_data/gateway_spec.rb]: not sure this test is required anymore

File: spec/active_merchant/billing/first_data/gateway_spec.rb [hunk 4]
```diff
@@ -230,75 +247,75 @@
...
 
       enter_credit_card_data(@trans_id, @visa_card_params)
 
       VCR.use_cassette('remote_11_result_ok') do
         response = @gateway.result(@trans_id, :client_ip_addr => @valid_ip)
-        response[:result].should == "OK"
-        response[:result_code].should == "000"
+        puts "Response"
+        puts response.inspect
+        puts response.class
```
Comment (ebeigarts) [inline spec/active_merchant/billing/first_data/gateway_spec.rb]: No puts please :)

File: lib/active_merchant/billing/first_data_response.rb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      def _3d_secure
+        parameters[:'3dsecure']
+      end
+
+      def result_message
+        result_text
+      end
+
+      def result_text
+        RESPONSE_CODES[result_code]
```
Comment (ebeigarts) [inline lib/active_merchant/billing/first_data_response.rb]: it is already set in initialzier as `message`

```
def result_text
  message
end
```

**Comment by ebeigarts** [comment]
Also tests are failing :(

**Comment by Sacristan** [comment]
@ebeigarts im back into this. Ill try fixing all the problems and then merge in Your master changes

**Comment by Sacristan** [comment]
Merged in master. Please evaluate code

**Comment by ebeigarts** [comment]
@Sacristan thnx, merged

**Comment by Sacristan** [comment]
@ebeigarts Can You make a release ?

**Comment by ebeigarts** [comment]
Released v1.3.0, let me know it there are issues.

**Pull request merged**